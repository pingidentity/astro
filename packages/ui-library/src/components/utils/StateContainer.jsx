import React from "react";
import PropTypes from "prop-types";
import Utils from "../../util/Utils";

const mapNothing = () => ({});

const defaultGetPassedProps = props => props;
const containerGetPassedProps = props => props.passedProps;

export const createProgressiveState = stateDefs => ({
    passedProps,
    initialState,
    boundSetState,
    getPassedProps = defaultGetPassedProps,
}) => {
    const isControlled = name => passedProps[name] === undefined;

    // figure out which props will be controlled and create initial state for those
    const pState = stateDefs.reduce((state, { name, initial }) => {
        if (isControlled(name)) {
            return {
                ...state,
                [name]: initialState[name] === undefined ? initial : initialState[name],
                // the initialState prop should override the defined initial value
                // but only if this is a controlled prop
            };
        }
        return state;
    }, {});

    // make sure the keys provided in initialState are actually being controlled
    if (!Utils.isProduction()) {
        Object.keys(initialState).forEach(key => {
            if (pState[key] === undefined) {
                console.warn(
                    `Initial state provided for prop "${key}" that is not being controlled by the component`
                );
            }
        });
    }

    // generate custom callbacks with transform functions
    const makeTransformCallbacks = (name, callbacks = []) =>
        callbacks.reduce(
            (transformCallbacks, { name: cbName, transform, passTransformedValue = false, }) => ({
                ...transformCallbacks,
                [cbName]: (value, e) => {
                    // set the state with the transformed value
                    boundSetState((state, props) => {
                        const transformedValue = transform(value, state[name], getPassedProps(props));

                        if (passedProps[cbName]) {
                            passedProps[cbName](passTransformedValue ? transformedValue : value, e);
                        }

                        return { [name]: transformedValue };
                    });
                }
            }), {}
        );

    // get the callbacks for a single state definition
    const makeCallbacksFromDef = ({ name, setter, callbacks }) => {
        // add any custom callbacks
        const customTransformCallbacks = makeTransformCallbacks(name, callbacks);

        // return with setter if defined
        return setter ? {
            [setter]: (value, e) => {
                boundSetState({ [name]: value });
                // if a callback is provided, we'll still execute it
                if (passedProps[setter]) {
                    passedProps[setter](value, e);
                }
            },
            ...customTransformCallbacks
        } : customTransformCallbacks;
    };

    // create all the callbacks
    const pCallbacks = stateDefs.reduce((returnCallbacks, def) => {
        if (!isControlled(def.name)) {
            return returnCallbacks;
        }
        return { ...returnCallbacks, ...makeCallbacksFromDef(def) };
    }, {});

    return [pState, pCallbacks];
};

class StateContainer extends React.Component {
    static propTypes = {
        stateDefs: PropTypes.arrayOf(
            PropTypes.shape({
                name: PropTypes.string.isRequired, // name of the props
                initial: PropTypes.any.isRequired, // starting value
                setter: PropTypes.string, // name of the callback in the target component
                callbacks: PropTypes.arrayOf( // when you need something different than a simple setter
                    PropTypes.shape({
                        name: PropTypes.string,
                        // Decides whether to pass the transformed value to the callback or just the value that
                        // the base component passes.
                        passTransformedValue: PropTypes.bool,
                        transform: PropTypes.func, // this function will be applied to the value and current state
                    })
                ),
            })
        ),
        initialState: PropTypes.object,
        passedProps: PropTypes.object,
        mapToProps: PropTypes.func,
    };

    static defaultProps = {
        stateDefs: [],
        initialState: {},
        passedProps: {},
        mapToProps: mapNothing,
    };

    constructor({ stateDefs, initialState, passedProps }) {
        // the constructor has two main goals
        // 1) populate the state for any props that will be controlled
        // 2) create callback functions for those props
        super();

        const [ state, callbacks ] = createProgressiveState(stateDefs)({
            initialState, passedProps, boundSetState: this.setState.bind(this), getPassedProps: containerGetPassedProps
        });

        this.state = state;
        this.callbacks = callbacks;
    }

    render() {
        const { children, mapToProps, passedProps } = this.props;

        // alert dev if there are state/prop conflicts
        if (!Utils.isProduction()) {
            Object.keys(this.state).forEach(key => {
                if (passedProps[key] !== undefined) {
                    console.warn(
                        `The prop "${key}" is being controlled by the component because it was not initially set. ` +
                        `The newly-received value will be ignored.`
                    );
                }
            });
        }

        return children({ ...passedProps, ...this.callbacks, ...this.state, ...mapToProps(this.state, passedProps) });
    }
}

export const inStateContainer = (stateDefs, mapToProps) => WrappedComponent => ({ initialState, ...props }) => (
    <StateContainer stateDefs={stateDefs} mapToProps={mapToProps} initialState={initialState} passedProps={props}>
        {containerProps => <WrappedComponent {...containerProps} />}
    </StateContainer>
);

export const toggleTransform = (value, current) => (!current);

export default StateContainer;