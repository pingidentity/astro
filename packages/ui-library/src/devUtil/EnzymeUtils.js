import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
