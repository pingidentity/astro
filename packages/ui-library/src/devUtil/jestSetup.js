
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
