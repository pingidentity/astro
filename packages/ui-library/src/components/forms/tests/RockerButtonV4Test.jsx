import { mount } from "enzyme";
import StateContainer from "../../utils/StateContainer";

import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";

describe("RockerButton v4", function () {

    const React = require("react"),
        ReactDOM = require("react-dom"),
        ReactTestUtils = require("react-dom/test-utils"),
        TestUtils = require("../../../testutil/TestUtils"),
        RockerButton = require("../RockerButton"),
        _ = require("underscore");

    const labelsArray = ["A", "B", "C"];

    function getComponent (opts) {
        opts = _.defaults(opts || {}, {
            onValueChange: jest.fn(),
            labels: labelsArray,
            className: "myRocker",
        });

        return ReactTestUtils.renderIntoDocument(<RockerButton {...opts} />);
    }

    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <RockerButton
                onValueChange={jest.fn()}
                labels={labelsArray}
            />
        );
    });

    it("data-id's don't change with helpText", () => {
        mountSnapshotDataIds(
            <RockerButton
                onValueChange={jest.fn()}
                labels={labelsArray}
                helpText="PLEASE HELP"
            />
        );
    });

    it("stateless: will render component with data-id", function () {
        const component = getComponent({
            "data-id": "myRockerButton"
        });

        const rocker = TestUtils.findRenderedDOMNodeWithDataId(component, "myRockerButton");

        expect(rocker).toBeDefined();
    });

    it("stateful: will render component with default data-id", function () {
        const component = getComponent();

        const rocker = TestUtils.findRenderedDOMNodeWithDataId(component, "rocker-button");

        expect(rocker).toBeDefined();
    });

    it("stateless: will trigger onValueChange callback when selection changes", function () {
        const component = getComponent();
        const labels = TestUtils.scryRenderedDOMNodesWithTag(component, "button");
        const testIndex = 2;

        ReactTestUtils.Simulate.click(labels[2], {});
        expect(component.props.onValueChange).toBeCalled();
        expect(component.props.onValueChange.mock.calls[0][0]).toEqual(
            { label: labelsArray[testIndex], index: testIndex }
        );
    });

    it("stateful: will trigger onValueChange callback when selection changes", function () {
        const callback = jest.fn();
        const component = ReactTestUtils.renderIntoDocument(
            <RockerButton labels={labelsArray} onValueChange={callback} />
        );
        const labels = TestUtils.scryRenderedDOMNodesWithTag(component, "button");
        const testIndex = 2;

        ReactTestUtils.Simulate.click(labels[testIndex], {});
        expect(component.props.onValueChange).toBeCalled();
        expect(component.props.onValueChange.mock.calls[0][0]).toEqual(
            { label: labelsArray[testIndex], index: testIndex }
        );
    });

    it("steteless: will not trigger callbacks if not given", function () {
        const callback = jest.fn();
        const component = ReactTestUtils.renderIntoDocument(
            <RockerButton labels={labelsArray} />);
        const labels = TestUtils.scryRenderedDOMNodesWithTag(component, "button");

        ReactTestUtils.Simulate.click(labels[2], {});
        expect(callback).not.toBeCalled();
    });

    it("stateful: will not trigger callbacks if not given", function () {
        const callback = jest.fn();
        const component = ReactTestUtils.renderIntoDocument(
            <RockerButton labels={labelsArray} />);
        const labels = TestUtils.scryRenderedDOMNodesWithTag(component, "button");

        ReactTestUtils.Simulate.click(labels[2], {});
        expect(callback).not.toBeCalled();
    });

    it("stateless: will accept index of selected item", function () {
        const component = getComponent({ selectedIndex: 1 });
        const node = TestUtils.findRenderedDOMNodeWithDataId(component, "rocker-button");

        expect(node.getAttribute("class").match("sel-1")).toBeTruthy();
    });

    it("will select initial index", function () {
        const component = getComponent({ selectedIndex: 2 });
        const node = TestUtils.findRenderedDOMNodeWithDataId(component, "rocker-button");

        expect(node.getAttribute("class").match("sel-2")).toBeTruthy();
    });

    it("will render labels rocker button list", function () {
        const component = getComponent();

        const labels = TestUtils.scryRenderedDOMNodesWithTag(component, "button");

        //make sure 3 labels in a list
        expect(labels.length).toBe(3);

        expect(labels[0].innerHTML).toBe("A");
        expect(labels[1].innerHTML).toBe("B");
        expect(labels[2].innerHTML).toBe("C");
    });

    it("disabled will not do anything for button clicks", function () {
        const callback = jest.fn(),
            component = ReactTestUtils.renderIntoDocument(
                <RockerButton disabled={true} labels={labelsArray} />),
            labels = TestUtils.scryRenderedDOMNodesWithTag(component, "button"),
            container = ReactDOM.findDOMNode(component);

        expect(component).toBeTruthy();
        expect(container.getAttribute("class")).toContain("disabled");

        ReactTestUtils.Simulate.click(labels[2], {});
        expect(callback).not.toBeCalled();
    });

    it("component", () => {
        const modal = mount(
            <RockerButton
                labels={labelsArray}
            />
        );
        expect(modal.find(StateContainer).exists()).toBeTruthy();

    });

    it("will trigger onValueChange callback when selection changes", function () {
        const callback = jest.fn();
        const component = ReactTestUtils.renderIntoDocument(
            <RockerButton labels={labelsArray} onValueChange={callback} />
        );
        const labels = TestUtils.scryRenderedDOMNodesWithTag(component, "button");
        const testIndex = 2;

        ReactTestUtils.Simulate.click(labels[testIndex], {});
        expect(component.props.onValueChange).toBeCalledWith(
            { label: labelsArray[testIndex], index: testIndex },
            undefined);
    });

    it("renders the selected class when the label is an object", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <RockerButton labels={[
                {
                    id: 7
                },
                {
                    id: 8
                }
            ]} selectedIndex={7} />
        );

        expect(TestUtils.findRenderedDOMNodeWithClass(component, "sel-0"));
    });

    it("renders the selected class when the label is an object and nothing selected", function() {
        const component = ReactTestUtils.renderIntoDocument(
            <RockerButton labels={[
                {
                    id: 7
                },
                {
                    id: 8
                }
            ]} />
        );

        expect(TestUtils.findRenderedDOMNodeWithClass(component, "sel-0"));
    });

});