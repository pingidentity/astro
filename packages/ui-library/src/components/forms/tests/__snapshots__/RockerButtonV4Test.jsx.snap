// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RockerButton v4 data-id's don't change 1`] = `
Array [
  "rocker-button",
  "rocker1",
  "icon",
  "icon-graphic",
  "icon-content",
  "rocker2",
  "rocker3",
  "rocker4",
]
`;

exports[`RockerButton v4 data-id's don't change with helpText 1`] = `
Array [
  "rocker-button",
  "rocker-label_a",
  "rocker-label_b",
  "rocker-label_c",
]
`;
